apollo-router/src/plugins/authentication.rs

```diff
@@ -0,0 +1,901 @@
+                            return failure_message(
+                                request.context,
+                                format!("Could not retrieve JWKS set: {e}"),
+                                StatusCode::INTERNAL_SERVER_ERROR, // XXX: Best error?
+                            );
+                        }
+                    };
+
+                    let jwks = match jwks_opt {
+                        Some(k) => k,
+                        None => {
+                            err_cleanup(false); // b.
+                            return failure_message(
+                                request.context,
+                                "Could not find JWKS set at the configured location".to_string(),
+                                StatusCode::INTERNAL_SERVER_ERROR,
+                            );
+                        }
+                    };
+
+                    // Now let's try to validate our token
+                    match jwks.find(&kid) {
+                        Some(jwk) => {
+                            let decoding_key = match DecodingKey::from_jwk(jwk) {
+                                Ok(k) => k,
+                                Err(e) => {
+                                    return failure_message(
+                                        request.context,
+                                        format!("Could not create decoding key: {}", e),
+                                        StatusCode::INTERNAL_SERVER_ERROR,
+                                    );
+                                }
+                            };
+
+                            let algorithm = match jwk.common.algorithm {
+                                Some(a) => a,
+                                None => {
+                                    return failure_message(
+                                        request.context,
+                                        "Jwk does not contain an algorithm".to_string(),
+                                        StatusCode::INTERNAL_SERVER_ERROR,
+                                    );
+                                }
+                            };
+
+                            let validation = Validation::new(algorithm);
+
+                            let token_data = match decode::<serde_json::Value>(
+                                jwt,
+                                &decoding_key,
+                                &validation,
+                            ) {
+                                Ok(v) => v,
+                                Err(e) => {
+                                    return failure_message(
+                                        request.context,
+                                        format!("Could not create decode JWT: {}", e),
+                                        StatusCode::UNAUTHORIZED,
+                                    );
+                                }
+                            };
```

Comment:
You have a lot of this pattern in this file, maybe we can create a macro to handle Option and Result ?

Reply:
I could do. It's a good suggestion. I'll do it if I find some spare time. Otherwise, maybe a follow-up.

Reply:
another approach here would be to make it as a function that returns a `Result<serde_json::Value, (String, StatusCode)>` and then have a wrapper that calls that function, and calls `failure_message` on the err case